docs/tutorials/linkage_example_verknupfung.qmd

```diff
@@ -0,0 +1,175 @@
+
+### Creating and assigning a feature factory
+
+The next step is to decide how to process each of the columns in our datasets.
+
+To do this, we define a feature factory that maps column types to feature
+generation functions, and a column specification for each dataset mapping our
+columns to column types in the factory.
+
+```{python}
+from pprl.embedder import features
+from functools import partial
+
+factory = dict(
+    name=features.gen_name_features,
+    sex=features.gen_sex_features,
+    misc=features.gen_misc_features,
+    dob=features.gen_dateofbirth_features,
+    instrument=partial(features.gen_misc_shingled_features,label='instrument')
+)
+spec1 = dict(
+    first_name="name",
+    last_name="name",
+    gender="sex",
+    instrument="instrument",
+    date_of_birth="dob",
+)
+spec2 = dict(name="name", sex="sex", main_instrument="instrument",birth_date="dob")
+```
+
+> [!TIP]
+> The feature generation functions, `features.gen_XXX_features` have sensible default parameters, but sometimes have to be passed in to the feature factory with different parameters, such as to set a feature label in the example above.
+> There are two ways to achieve this. Either use `functools.partial` to set parameters (as above), or pass keyword arguments as a dictionary of dictionaries to the Embedder as `ff_args`.
+
+### Embedding the data
+
+With our specifications sorted out, we can get to creating our Bloom filter
+embedding. Before doing so, we need to decide on two parameters: the size of
+the filter and the number of hashes. By default, these are `1024` and `2`,
+respectively.
+
+Once we've decided, we can create our `Embedder` instance and use it to embed
+our data with their column specifications.
+
+```{python}
+from pprl.embedder.embedder import Embedder
+
+embedder = Embedder(factory, bf_size=1024, num_hashes=2)
+edf1 = embedder.embed(df1, colspec=spec1, update_thresholds=True)
+edf2 = embedder.embed(df2, colspec=spec2, update_thresholds=True)
+```
+
+If we take a look at one of these embedded datasets, we can see that it has a
+whole bunch of new columns. There is a `_features` column for each of the
+original columns containing their pre-embedding string features, and there's an `all_features` column that combines the features. Then there are
+three additional columns: `bf_indices`, `bf_norms` and `thresholds`.
+
+```{python}
+edf1.columns
+```
+
+The `bf_indices` column contains the Bloom filters, represented compactly as a list of non-zero indices for each record. The `bf_norms` column contains the norm of each Bloom filter with respect to the Soft Cosine Measure (SCM) matrix. In this case since we are using an untrained model, the SCM matrix is an identity matrix, and the norm is just the euclidean norm of the Bloom filter represented as a binary vector, which is equal to `np.sqrt(len(bf_indices[i]))` for record `i`. The norm is used to scale the similarity measures so that they take values between -1 and 1.
```

Comment:
```suggestion
The `bf_indices` column contains the Bloom filters, represented compactly as a list of non-zero indices for each record. The `bf_norms` column contains the norm of each Bloom filter with respect to the Soft Cosine Measure (SCM) matrix. In this case since we are using an untrained model, the SCM matrix is an identity matrix, and the norm is just the Euclidean norm of the Bloom filter represented as a binary vector, which is equal to `np.sqrt(len(bf_indices[i]))` for record `i`. The norm is used to scale the similarity measures so that they take values between -1 and 1.
```